PngPlugin: help message if dependency is not compiled

psychocoderHPC opened the pull request:
If `pngwriter` is not available at compile time, the plugin `PngPlugin` register a help with information why the plugin can't use at runtime.

- PngCreator: add "tpp` file for `operator()`
- PngPlugin: add help message if `pngwriter` is not avail


This is the first plugin which full fill #1700


```bash
./picongpu --help
# ...
PositionsParticles: write position of one particle of a species to std::cout:
  --i_position.period arg               enable analyser [for each n-th step]

PngPlugin: create pngs of a species and fields:
  --i_png                               plugin disabled [compiled without 
                                        dependency pngwriter]
# ...
PositionsParticles: write position of one particle of a species to std::cout:
  --e_position.period arg               enable analyser [for each n-th step]

PngPlugin: create pngs of a species and fields:
  --e_png                               plugin disabled [compiled without 
                                        dependency pngwriter]
# ...
```

psychocoderHPC commented:
The only issu is that some plugin were generated for each species type and therefore write the message with missing dependency multiple time (see description example).
We can avoid this if the png plugin itself generate an instance for each species.

**We can solve each issue with a new abstraction :-)**

ax3l commented:
that's ok for me, it's as verbose as it is when enabled...

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.hpp:
```diff
@@ -136,75 +128,25 @@ namespace picongpu
+} /* namespace picongpu */
+
+#if(PIC_ENABLE_PNG==1)
+#include "plugins/output/images/PngCreator.tpp"
```
ah ah
```C++
#if( PIC_ENABLE_PNG == 1 )
#    include "plugins/output/images/PngCreator.tpp"
#else
```

;)

ax3l commented on src/picongpu/include/plugins/PngPlugin.hpp:
```diff
@@ -71,11 +71,16 @@ namespace picongpu
 
         void pluginRegisterHelp(po::options_description& desc)
         {
+#if(PIC_ENABLE_PNG==1)
```
nobody does spaces :( `#if( PIC_ENABLE_PNG == 1 )`

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.hpp:
```diff
@@ -136,75 +128,25 @@ namespace picongpu
+     */
     template<class Box>
     inline void PngCreator::createImage(
                                         const Box data,
```
Y u still align so badly? ;)

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.hpp:
```diff
@@ -136,75 +128,25 @@ namespace picongpu
+{
+    /** empty implementation if PIConGPU is compiled without PNGwriter
+     *
+     * If called this method will always fails with an exception at runtime.
```
typo: If called this method always fail with ...

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.tpp:
```diff
@@ -0,0 +1,110 @@
+{
+    template<class Box>
+    inline void PngCreator::createImage(
+                                        const Box data,
```
Y u still align so badly? ;)

psychocoderHPC commented:
I updated this style issues.
I moved the assert for the case that `PNGwriter` is not loaded but the png creation method is called from `PngCreator.hpp` to `PngCreator.tpp`. This avoid copy past of the signature of `createImage(...)`

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.tpp:
```diff
@@ -0,0 +1,120 @@
+{
+    template< class Box >
+    inline void PngCreator::createImage(
+        const Box data,
```
this will cause unused variable warnings when PNGwriter is not available, *even if the user does not plan to use it*.

ax3l replied:
either use
```C++
inline void PngCreator::createImage(
    const Box /*data*/,
    const Size2D /*size*/,
    const MessageHeader /*header/*
)
```
there or avoid calling that interface at RT.

ax3l replied:
offline discussion: self-implement boost `unused`

psychocoderHPC replied:
I realized that `createImage` is a host only function without any prefix, therefore we should not use an self implemented ignore function. For this case we should use `boost::ignore_unused`.
Never the less we need a equivalent function with the prefix `__host__ __device__` for other code parts.

BenjaminW3 replied:
You could try to add `BOOST_GPU_ENABLED` to [`boost::ignore_unused`](https://github.com/boostorg/core/blob/develop/include/boost/core/ignore_unused.hpp) like:
```
template <typename... Ts>
BOOST_FORCEINLINE BOOST_GPU_ENABLED BOOST_CXX14_CONSTEXPR void ignore_unused(Ts const& ...)
{}
```
If this would work and could be integrated into boost, alpaka would be happy too 😄

BenjaminW3 replied:
However, I do not currently have access to a Linux Box with an easily customizable Boost and CUDA, so I can not verify this beforehand. Could someone do this for me?

psychocoderHPC commented:
updated

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.hpp:
```diff
@@ -136,75 +127,6 @@ namespace picongpu
-
 } /* namespace picongpu */
+
+#include "plugins/output/images/PngCreator.tpp"
```
Can't we make this a regular `.cpp` file? `PngCreator` is template-free :)

ax3l replied:
Can't we make this a regular `.cpp` file? `PngCreator` is template-free :)

psychocoderHPC replied:
The method `createImage` used a template `Box`, therefore it is not possible.

ax3l replied:
:'(

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.tpp:
```diff
@@ -0,0 +1,122 @@
+            png.scale_kxky( scale_x, scale_y );
+
+        // add some meta information
+        //header.writeToConsole( std::cout );
```
got ya! dead code :)

ax3l commented on src/picongpu/include/plugins/output/images/PngCreator.tpp:
```diff
@@ -0,0 +1,122 @@
+         */
+        png.setcompressionlevel( 1 );
+
+        //PngWriter coordinate system begin with 1,1
```
ok, in that case lets add a space after `//` to be consistent
